libsql-wal/src/wal.rs

```diff
@@ -0,0 +1,303 @@
+
+#[derive(Clone)]
+pub struct LibsqlWalManager<FS: FileSystem> {
+    pub registry: Arc<WalRegistry<FS>>,
+    pub next_conn_id: Arc<AtomicU64>,
+}
+
+pub struct LibsqlWal<FS: FileSystem> {
+    last_read_frame_no: Option<u64>,
+    tx: Option<Transaction<FS::File>>,
+    shared: Arc<SharedWal<FS>>,
+    conn_id: u64,
+}
+
+impl<FS: FileSystem> WalManager for LibsqlWalManager<FS> {
+    type Wal = LibsqlWal<FS>;
+
+    fn use_shared_memory(&self) -> bool {
+        false
+    }
+
+    fn open(
+        &self,
+        _vfs: &mut libsql_sys::wal::Vfs,
+        _file: &mut libsql_sys::wal::Sqlite3File,
+        _no_shm_mode: std::ffi::c_int,
+        _max_log_size: i64,
+        db_path: &std::ffi::CStr,
+    ) -> libsql_sys::wal::Result<Self::Wal> {
+        let db_path = OsStr::from_bytes(&db_path.to_bytes());
+        let shared = self
+            .registry
+            .clone()
+            .open(db_path.as_ref())
+            .map_err(Into::into)?;
+        let conn_id = self
+            .next_conn_id
+            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+        Ok(LibsqlWal {
+            last_read_frame_no: None,
+            tx: None,
+            shared,
+            conn_id,
+        })
+    }
+
+    fn close(
+        &self,
+        wal: &mut Self::Wal,
+        _db: &mut libsql_sys::wal::Sqlite3Db,
+        _sync_flags: std::ffi::c_int,
+        _scratch: Option<&mut [u8]>,
+    ) -> libsql_sys::wal::Result<()> {
+        wal.end_read_txn();
+        Ok(())
+    }
+
+    fn destroy_log(
+        &self,
+        _vfs: &mut libsql_sys::wal::Vfs,
+        _db_path: &std::ffi::CStr,
+    ) -> libsql_sys::wal::Result<()> {
+        Ok(())
+    }
+
+    fn log_exists(
+        &self,
+        _vfs: &mut libsql_sys::wal::Vfs,
+        _db_path: &std::ffi::CStr,
+    ) -> libsql_sys::wal::Result<bool> {
+        Ok(true)
+    }
+
+    fn destroy(self)
+    where
+        Self: Sized,
+    {
+    }
+}
+
+impl<FS: FileSystem> Wal for LibsqlWal<FS> {
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn limit(&mut self, _size: i64) {}
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id, ns = self.shared.namespace().as_str()))]
+    fn begin_read_txn(&mut self) -> libsql_sys::wal::Result<bool> {
+        tracing::trace!("begin read");
+        let tx = self.shared.begin_read(self.conn_id);
+        let invalidate_cache = self
+            .last_read_frame_no
+            .map(|idx| tx.max_frame_no != idx)
+            .unwrap_or(true);
+        self.last_read_frame_no = Some(tx.max_frame_no);
+        self.tx = Some(Transaction::Read(tx));
+
+        tracing::debug!(invalidate_cache, "read started");
+        Ok(invalidate_cache)
+    }
+
+    #[tracing::instrument(skip_all, fields(id = self.conn_id))]
+    fn end_read_txn(&mut self) {
+        match self.tx.take() {
+            Some(Transaction::Read(tx)) => tx,
+            Some(Transaction::Write(tx)) => tx.downgrade(),
```

Comment:
in which situation a write transaction can call `end_read_txn`?

Reply:
Usually, SQLite will end the write and then end the read, but in OG SQLite WAL implementation, ending a read transaction also ends the write

Reply:
Although ending a write txn does not end a read txn